### PR TITLE
fix engravings that have gone out of sync

### PR DESCRIPTION
**Pull request by myk002**
Fixes: https://github.com/DFHack/dfhack/issues/4421

### CODE REVIEW

**Comment by c3r341** [comment]
pre-commit.ci autofix

File: engravings.lua [hunk 1]
```diff
@@ -0,0 +1,35 @@
+--DFHack:fix/engravings.lua
+--4/23/24
+--@lord_cereal
```
Comment (myk002) [inline engravings.lua:3]: you can remove these -- they are recorded in the git commit history and don't need to be duplicated here

File: engravings.lua
Comment (myk002) [inline engravings.lua:1]: indenting of the code needs cleanup. four spaces per indent, please

File: engravings.lua [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	end
+end
+
+--loop runs through list of all engravings checking each using is_good_engraving and if bad gets deleted
+local cleanup = 0
+local engravings = df.global.world.engravings
+for index = #engravings-1,0,-1 do
+    local engraving = engravings[index]
+	if not is_good_engraving(engraving) then
+        df.global.world.engravings:erase(index)
```
Comment (myk002) [inline engravings.lua:30]:
```suggestion
        engravings:erase(index)

```

File: engravings.lua
Comment (myk002) [inline engravings.lua:1]: Could you move the file to `fix/engravings.lua` so it's in the proper directory?

Comment (myk002) [inline engravings.lua:1]: needs documentation in `docs/fix/engravings.rst` -- you can copy the formatting and most of the content from one of the similar `fix/` scripts

Comment (myk002) [inline engravings.lua:1]: needs a line added to `internal/control-panel/registry.lua`. copy a line from a similar `fix/` script and add it in alphabetical order. `fix/stuck-worship` is probably a good one to copy from.

**Comment by myk002** [comment]
pre-commit.ci autofix

File: fix/engravings.lua [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+local engravings = df.global.world.engravings
+for index = #engravings-1,0,-1 do
+    local engraving = engravings[index]
+    if not is_good_engraving(engraving) then
+        engravings:erase(index)
+        engraving:delete()
+        cleanup = cleanup + 1
+    end
+end
+print(('%d bad engraving(s) fixed.'):format(cleanup))
```
Comment (myk002) [inline fix/engravings.lua:31]:
```suggestion
if not quiet or cleanup > 0 then
    print(('%d bad engraving(s) fixed.'):format(cleanup))
end
```

File: fix/engravings.lua [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    if tileattrs.shape == df.tiletype_shape.FLOOR then
+        return engraving.flags.floor
+    end	
+	
+    if tileattrs.shape == df.tiletype_shape.WALL then
+        return not engraving.flags.floor
+    end
+end
+
+--loop runs through list of all engravings checking each using is_good_engraving and if bad gets deleted
```
Comment (myk002) [inline fix/engravings.lua:20]:
```suggestion
local help = false
local quiet = false
local positionals = argparse.processArgsGetopt(args, {
    {'h', 'help', handler=function() help = true end},
    {'q', 'quiet', handler=function() quiet = true end},
})

if help or positionals[1] == 'help' then
    print(dfhack.script_help())
    return
end

--loop runs through list of all engravings checking each using is_good_engraving and if bad gets deleted
```

File: fix/engravings.lua [hunk 5]
```diff
@@ -0,0 +1,31 @@
+--function checks tiletype for attributes and returns true or false depending on if its engraving location is correct
```
Comment (myk002) [inline fix/engravings.lua:1]:
```suggestion
local argparse = require('argparse')

--function checks tiletype for attributes and returns true or false depending on if its engraving location is correct
```

File: internal/control-panel/registry.lua [hunk 6]
```diff
@@ -72,6 +72,8 @@ COMMANDS_BY_IDX = {
     {command='fix/empty-wheelbarrows', group='bugfix', mode='repeat', default=true,
         desc='Make abandoned full wheelbarrows usable again.',
         params={'--time', '1', '--timeUnits', 'days', '--command', '[', 'fix/empty-wheelbarrows', '-q', ']'}},
+    {command='fix/engravings', group='bugfix', mode='repeat', default=true,
+        params={'--time', '1', '--timeUnits', 'days', '--command', '[', 'fix/engravings', ']'}},
```
Comment (myk002) [inline internal/control-panel/registry.lua:76]:
```suggestion
        params={'--time', '1', '--timeUnits', 'days', '--command', '[', 'fix/engravings', '-q', ']'}},
```

**Review by myk002 — COMMENTED** [review]
added some boilerplate so that the code runs silently if no changes are made (otherwise the console gets cluttered with the scheduled daily run)

File: docs/fix/engravings.rst [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+When placing a new wall or new floor down where a previous engraved tiletype was, the tile may be corrupted and unengravable. 
+This fix removes corrupted engravings from those tiles automatically so those tiletypes may be engraved again.
+
+Usage
+-----
+
+::
+
+    fix/engravings
```
Comment (myk002) [inline docs/fix/engravings.rst:16]:
```suggestion
    fix/engravings [<options>]

Options
-------

``-q``, ``--quiet``
    Only output status when something was actually fixed.
```

File: docs/fix/engravings.rst [hunk 8]
```diff
@@ -0,0 +1,16 @@
+fix/engravings
+=============
+
+.. dfhack-tool::
+    :summary: Fixes unengravable corrupted tiles so they are able to be engraved.
+    :tags: fort bugfix
+
+When placing a new wall or new floor down where a previous engraved tiletype was, the tile may be corrupted and unengravable. 
+This fix removes corrupted engravings from those tiles automatically so those tiletypes may be engraved again.
```
Comment (myk002) [inline docs/fix/engravings.rst:9]:
```suggestion
When constructing a new wall or new floor over a previously engraved tile, the tile may become corrupted and unengravable. 
This fix detects the problem and resets the state of those tiles so they may be engraved again.

```

File: docs/fix/engravings.rst [hunk 9]
```diff
@@ -0,0 +1,16 @@
+fix/engravings
+=============
```
Comment (myk002) [inline docs/fix/engravings.rst:2]:
```suggestion
==============
```

File: docs/fix/engravings.lua
Comment (myk002) [inline docs/fix/engravings.lua:1]: extra file?

**Review by myk002 — COMMENTED** [review]
also, a line needs to be added to the `changelog.txt` file, "Future" section, "New tools" subheader

File: docs/fix/engravings.lua
Comment (c3r341) [reply docs/fix/engravings.lua:1]: ive deleted it 3 times? weird

**Comment by myk002** [comment]
pre-commit.ci autofix

File: changelog.txt [hunk 10]
```diff
@@ -30,6 +30,7 @@ Template for new versions:
 - Updated for adventure mode: `gui/sandbox`, `gui/create-item`
 - `markdown`: (reinstated) export unit or item descriptions to a markdown text file
 - `adaptation`: (reinstated) inspect or set unit cave adaptation levels
+- `engravings`: remove corrupt engraving tiles
```
Comment (myk002) [inline changelog.txt:33]:
```suggestion
- `fix/engravings`: remove corrupt engraving tiles
```

File: engravings.lua
Comment (myk002) [inline engravings.lua:1]: extra copy of file

File: docs/fix/engravings.lua
Comment (myk002) [inline docs/fix/engravings.lua:1]: extra copy of file

**Review by myk002 — COMMENTED** [review]
once the two extra copies of engravings.lua are removed, this looks good to go!

- /fix/engravings.lua should stay (this is the correct one)
- /docs/fix/engravings.lua should be deleted
- /engravings.lua should be deleted

**Comment by myk002** [comment]
I think I can help fix the remaining issues in a separate branch. Hang tight.

**Pull request merged**